Content.Server/Atmos/Piping/Unary/EntitySystems/GasThermoMachineSystem.cs

```diff
@@ -58,26 +58,40 @@ private void OnThermoMachineUpdated(EntityUid uid, GasThermoMachineComponent the
             }
 
             float sign = Math.Sign(thermoMachine.Cp); // 1 if heater, -1 if freezer
-
-            // Implement hysteresis
-            float hystTemp = thermoMachine.TargetTemperature;
+            bool heater = sign > 0;
+            float targetTemp = thermoMachine.TargetTemperature;
+            float dTHyst = sign * thermoMachine.TemperatureTolerance;
+            float temp = inlet.Air.Temperature;
+
+            if (sign * temp >= sign * (targetTemp + dTHyst))
+                thermoMachine.HysteresisState = false;
+            if (sign * temp < sign * targetTemp)
+                thermoMachine.HysteresisState = true;
             if (thermoMachine.HysteresisState)
-            {
-                hystTemp += sign*thermoMachine.TemperatureTolerance;
-            }
+                targetTemp += dTHyst;
 
-            // Stop if we've reached our target temperature.
-            if (sign*inlet.Air.Temperature > sign*hystTemp) // inequality flips if freezer
+            float dT = targetTemp - temp;
+            if (sign != Math.Sign(dT))
```

Comment:
Is this the same as `if (thermoMachine.HysteresisState)`?

Reply:
it looks like it is, but this makes the intent more clear and avoids unforeseen issues

Reply:
What are the unforseen issues? Isn't `HysteresisState` now the same as "should the heater be on"? That can be noted in the comments.

Reply:
i guess i'll change it to check for !thermoMachine.HysteresisState then